### PR TITLE
[Update] Firebase for iOS v5.4.0

### PR DESCRIPTION
**Pull request by SotoiGhost**
Components updated to:

* [x] Firebase Core v5.0.0
* [x] Firebase InstanceID v3.0.0
* [x] Firebase Analytics v5.0.0
* [x] Firebase DynamicLinks v3.0.0
* [x] Firebase CloudMessaging v3.0.0
* [x] Google SignIn v4.1.2
* [x] Firebase AdMob v7.30.0
* [x] Firebase Auth v5.0.0
* [x] Firebase Invites v3.0.0
* [x] Firebase Database 5.0.0
* [x] Firebase Storage v3.0.0
* [x] Firebase CloudFirestore v0.12.0
* [x] Firebase ABTesting v2.0.0
* [x] Firebase RemoteConfig v3.0.0
* [x] Firebase PerformanceMonitoring v2.0.0
* [x] Firebase Crashlytics v3.10.3

### CODE REVIEW

**Comment by SotoiGhost** [comment]
@dalexsoto Mark the component name as you review it if everything seems ok. Thanks in advance!

* [ ] Firebase Core v5.0.0
* [ ] Firebase InstanceID v3.0.0
* [ ] Firebase Analytics v5.0.0
* [ ] Firebase AdMob v7.30.0
* [ ] Firebase DynamicLinks v3.0.0
* [ ] Firebase Invites v3.0.0
* [ ] Firebase PerformanceMonitoring 2.0.0
* [ ] Firebase Auth v5.0.0
* [ ] Firebase Database 5.0.0
* [ ] Firebase CloudFirestore v0.12.0
* [ ] Firebase Function v2.0.0
* [ ] Firebase CloudMessaging v3.0.0
* [ ] Firebase Storage v3.0.0

File: Firebase.Analytics/source/Firebase.Analytics/ApiDefinition.cs [hunk 1]
```diff
@@ -1,11 +1,13 @@
 ﻿using System;
+using System.Collections.Generic;
```
Comment (dalexsoto) [inline Firebase.Analytics/source/Firebase.Analytics/ApiDefinition.cs]: Why is this needed?

**Comment by dalexsoto** [comment]
LGTM so far, reviewed the following:

* [x] Firebase Core v5.0.0
* [x] Firebase InstanceID v3.0.0
* [x] Firebase Analytics v5.0.0
* [x] Firebase AdMob v7.30.0
* [x] Firebase DynamicLinks v3.0.0

**Comment by SotoiGhost** [comment]
@Redth As the same process that @dalexsoto and I are doing, If you want to review the PR while I'm updating it.

Things to review:

* Component version on **component.yaml**, **nuspec**, **targets** and **Readme** files
* **External-Dependency-Info** file update, if any

Mark the component name as you review it if everything seems ok. Thanks in advance!

* [ ] Firebase Core v5.0.0
* [ ] Firebase InstanceID v3.0.0
* [ ] Firebase Analytics v5.0.0
* [ ] Firebase AdMob v7.30.0
* [ ] Firebase DynamicLinks v3.0.0
* [ ] Firebase Invites v3.0.0 [NotReady]
* [ ] Firebase PerformanceMonitoring 2.0.0 [NotReady]
* [ ] Firebase Auth v5.0.0 [NotReady]
* [ ] Firebase Database 5.0.0 [NotReady]
* [ ] Firebase CloudFirestore v0.12.0 [NotReady]
* [ ] Firebase Function v2.0.0 [NotReady]
* [ ] Firebase CloudMessaging v3.0.0 [NotReady]
* [ ] Firebase Storage v3.0.0 [NotReady]

**Comment by dalexsoto** [comment]
LGTM 👍

**Comment by SotoiGhost** [comment]
@dalexsoto I added the Google Mobile Ads Consent SDK binding to [Firebase AdMob][1]. Can you, please, give a check again? Thanks in advance!

[1]: https://github.com/xamarin/GoogleApisForiOSComponents/pull/152/commits/b5769d619e454f67fe2bc280040f6d2e6877ff71#diff-e62775556e2070750942263534aacfbc

File: Google.MobileAds/source/Google.MobileAds/ApiDefinition.cs [hunk 2]
```diff
@@ -481,6 +481,10 @@ interface Request : INSCopying {
 		void SetLocation (string locationDescription);
 
 		[Export ("tagForChildDirectedTreatment:")]
+		void Tag (bool forChildDirectedTreatment);
+
+		[Obsolete ("Please, use Tag method instead. This will be removed in future versions.")]
+		[Wrap ("TagForChildDirectedTreatment (childDirectedTreatment)")]
```
Comment (dalexsoto) [inline Google.MobileAds/source/Google.MobileAds/ApiDefinition.cs]: This is wrong :) you are calling yourself in an infinite loop.

**Comment by SotoiGhost** [comment]
@dalexsoto [Firebase Auth][1] is ready for review. Thanks dude!

[1]: https://github.com/xamarin/GoogleApisForiOSComponents/pull/152/commits/6ce28b65431dc8a25eb61827b50c0af9458f280a#diff-5623da69229c4dc2352821c3e6771f58

**Comment by SotoiGhost** [comment]
@dalexsoto Firebase Invites is ready for review. Thanks dude!

File: Firebase.Invites/source/Firebase.Invites/ApiDefinition.cs [hunk 3]
```diff
@@ -29,9 +29,9 @@ interface IInviteDelegate
 	}
 
 	// @protocol FIRInviteDelegate <NSObject>
-	[Model]
+	[Model (AutoGeneratedName = true)]
 	[Protocol]
-	[BaseType (typeof (NSObject), Name = "FIRInviteDelegate")]
```
Comment (dalexsoto) [inline Firebase.Invites/source/Firebase.Invites/ApiDefinition.cs]: You removed the native name `FIRInviteDelegate` so the protocol won't be found at all

**Comment by SotoiGhost** [comment]
@dalexsoto [Firebase.CloudFirestore][1] is ready for review. Thanks bro!

[1]: https://github.com/xamarin/GoogleApisForiOSComponents/pull/152/commits/2f8040170c3337ca72bcd5386bdb52a49f66c9d2#diff-85afcd023282c2677199761efeafc9e3

**Comment by SotoiGhost** [comment]
@dalexsoto [Firebase.ABTesting][1] is ready for review. Danke!

[1]: https://github.com/xamarin/GoogleApisForiOSComponents/pull/152/commits/786687f43e832388c0cdf4ad5f227a01d92f7533#diff-ed240fc14f0e294a301d165fb1806dcb

**Comment by SotoiGhost** [comment]
@dalexsoto [Firebase.RemoteConfig] is ready for review. Danke again!

[1]: https://github.com/xamarin/GoogleApisForiOSComponents/pull/152/commits/6c1f9cd79be05074d78cae4e77d5d9916f412cbb#diff-5a98db42b0a9851996dae6c868775551

**Comment by SotoiGhost** [comment]
@dalexsoto [Firebase.PerformanceMonitoring][1] is ready for review. Thanks in advance!

[1]: https://github.com/xamarin/GoogleApisForiOSComponents/pull/152/commits/b38439c28fbf1db5098da9e97e5ef50391208759#diff-39108cf406b0b117c51b76d48d67bc00

File: Firebase.ABTesting/source/Firebase.ABTesting/ApiDefinition.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		// -(NSTimeInterval)latestExperimentStartTimestampBetweenTimestamp:(NSTimeInterval)timestamp andPayloads:(NSArray<NSData *> * _Nonnull)payloads;
+		[Export ("latestExperimentStartTimestampBetweenTimestamp:andPayloads:")]
+		double GetLatestExperimentStartTimestampBetweenTimestamp (double timestamp, NSData [] payloads);
+	}
+
+	// @interface FIRLifecycleEvents : NSObject
+	[BaseType (typeof (NSObject), Name = "FIRLifecycleEvents")]
+	interface LifecycleEvents {
+		// extern NSString *const _Nonnull FIRSetExperimentEventName;
+		[Field ("FIRSetExperimentEventName", "__Internal")]
```
Comment (dalexsoto) [inline Firebase.ABTesting/source/Firebase.ABTesting/ApiDefinition.cs]: I would bind this on its own static class

```csharp
[Static]
interface DefaultLifecycleEventNames {

	[Field ("FIRSetExperimentEventName", "__Internal")]
	NSString SetExperiment { get; }

	// extern NSString *const _Nonnull FIRActivateExperimentEventName;
	[Field ("FIRActivateExperimentEventName", "__Internal")]
	NSString ActivateExperiment { get; }

	// extern NSString *const _Nonnull FIRClearExperimentEventName;
	[Field ("FIRClearExperimentEventName", "__Internal")]
	NSString ClearExperiment { get; }

	// extern NSString *const _Nonnull FIRTimeoutExperimentEventName;
	[Field ("FIRTimeoutExperimentEventName", "__Internal")]
	NSString TimeoutExperiment { get; }

	// extern NSString *const _Nonnull FIRExpireExperimentEventName;
	[Field ("FIRExpireExperimentEventName", "__Internal")]
	NSString ExpireExperiment { get; }
}
```

and I would add an `Advice` attribute to point this out

```csharp
[Advice ("You can use the default event name 'DefaultLifecycleEventNames.SetExperiment'.")]
[Export ("setExperimentEventName")]
string SetExperimentEventName { get; set; }
```

File: Firebase.ABTesting/source/Firebase.ABTesting/ApiDefinition.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		[Field ("FIRTimeoutExperimentEventName", "__Internal")]
+		NSString DefaultTimeoutExperimentEventName { get; }
+
+		// extern NSString *const _Nonnull FIRExpireExperimentEventName;
+		[Field ("FIRExpireExperimentEventName", "__Internal")]
+		NSString DefaultExpireExperimentEventName { get; }
+
+		// @property (copy, nonatomic) NSString * _Nonnull setExperimentEventName;
+		[Export ("setExperimentEventName")]
+		string SetExperimentEventName { get; set; }
```
Comment (dalexsoto) [inline Firebase.ABTesting/source/Firebase.ABTesting/ApiDefinition.cs]: There is one reason you might want to use `NSString` here instead of `string`. Since there is a default value `FIRSetExperimentEventName` the native API might be doing a pointer check instead of a string contents check.

You will need to make a decision based on documentation and the most likely scenario. For example if this API does not change its behavior even if the default name is used then you might be able to keep `string` but if a special thing happens if you use the default string name then you might need to switch to `NSString`.

This rule apply to all string properties below inside of this class.

File: Firebase.RemoteConfig/source/Firebase.RemoteConfig/ApiDefinition.cs [hunk 6]
```diff
@@ -89,20 +90,22 @@ interface RemoteConfig
...
 		// -(BOOL)activateFetched;
 		[Export ("activateFetched")]
 		bool ActivateFetched ();
 
-        // -(FIRRemoteConfigValue * _Nonnull)objectForKeyedSubscript:(NSString * _Nonnull)key;
-        [Export("objectForKeyedSubscript:")]
-        RemoteConfigValue ObjectForKeyedSubscript(string key);
+		// -(FIRRemoteConfigValue * _Nonnull)objectForKeyedSubscript:(NSString * _Nonnull)key;
+		[Export("objectForKeyedSubscript:")]
+		RemoteConfigValue ObjectForKeyedSubscript(string key);
```
Comment (dalexsoto) [inline Firebase.RemoteConfig/source/Firebase.RemoteConfig/ApiDefinition.cs]: Nope. never expose this thing `ObjectForKeyedSubscript` or at least not named like that, you can use `GetObject` as name or you could expose a [C# indexer](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/indexers/) instead using this API internally.

**Comment by SotoiGhost** [comment]
@dalexsoto This is the new list that I will update. Hope, will be easier and quicker than the previous wave

* [ ] Firebase Core v5.0.5
* [ ] Firebase InstanceID v3.1.1
* [ ] Firebase Analytics v5.0.1
* [ ] Firebase DynamicLinks v3.0.1
* [ ] Firebase CloudMessaging v3.0.3
* [ ] Firebase AdMob v7.31.0
* [ ] Firebase Auth v5.0.2
* [ ] Firebase Database 5.0.2
* [ ] Firebase CloudFirestore v0.12.5
* [ ] Firebase PerformanceMonitoring v2.0.1

**Comment by SotoiGhost** [comment]
@dalexsoto  [Firebase.AdMob][1] and [Firebase.CloudFirestore][2] are ready for review. Thank you!

[1]: https://github.com/xamarin/GoogleApisForiOSComponents/pull/152/commits/5887b947715da8f9a79e196de748b5a4839560dc#diff-e62775556e2070750942263534aacfbc
[2]: https://github.com/xamarin/GoogleApisForiOSComponents/pull/152/commits/00f4093d85fa3bceb5ecff990349ae373b741b7c#diff-85afcd023282c2677199761efeafc9e3

File: Firebase.ABTesting/source/Firebase.ABTesting/ApiDefinition.cs [hunk 7]
```diff
@@ -19,47 +20,51 @@ interface ExperimentController {
...
-		[Export ("setExperimentEventName")]
-		string SetExperimentEventName { get; set; }
+		[Advice ("You can use the default event name 'DefaultLifecycleEventNames.SetExperiment'.")]
+		[Export ("setExperimentEventName", ArgumentSemantic = ArgumentSemantic.Copy)]
+		NSString SetExperimentEventName { get; set; }
 
 		// @property (copy, nonatomic) NSString * _Nonnull activateExperimentEventName;
-		[Export ("activateExperimentEventName")]
-		string ActivateExperimentEventName { get; set; }
+		[Export ("activateExperimentEventName", ArgumentSemantic = ArgumentSemantic.Copy)]
```
Comment (dalexsoto) [inline Firebase.ABTesting/source/Firebase.ABTesting/ApiDefinition.cs]: Advice: Add [Advice] 

/me giggles 

tenor

File: Firebase.ABTesting/nuget/Xamarin.Firebase.iOS.ABTesting.nuspec [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+        <dependency id="Xamarin.Firebase.iOS.Core" version="5.0.5.0" />
+      </group>
+    </dependencies>
+  </metadata>
+  <files>
+    <file src="output/Firebase.ABTesting.dll" target="lib/Xamarin.iOS10" />
+
+    <!-- Add targets file for external native reference download -->
+    <file src="source/Firebase.ABTesting/Firebase.ABTesting.targets" target="build/Xamarin.Firebase.iOS.ABTesting.targets" />
+  </files>
```
Comment (Redth) [inline Firebase.ABTesting/nuget/Xamarin.Firebase.iOS.ABTesting.nuspec]: Missing External dep info / third party notice file inclusion?

File: Google.MobileAds/External-Dependency-Info [hunk 9]
```diff
@@ -0,0 +1,219 @@
+THIRD-PARTY SOFTWARE NOTICES AND INFORMATION
+Do not translate or localize
+
+Xamarin Components for Firebase APIs Cloud Messaging iOS Library incorporates third party
```
Comment (Redth) [inline Google.MobileAds/External-Dependency-Info]: Copy paste? Should be `for Google Mobile Ads` ?

File: Firebase.ABTesting/nuget/Xamarin.Firebase.iOS.ABTesting.nuspec (hunk 8 above)
Comment (Redth) [reply Firebase.ABTesting/nuget/Xamarin.Firebase.iOS.ABTesting.nuspec]: Using XBD, not needed, nm.

File: Firebase.Analytics/nuget/Xamarin.Firebase.iOS.Analytics.nuspec [hunk 10]
```diff
@@ -14,9 +14,9 @@
...
       <group targetFramework="Xamarin.iOS10">
-        <dependency id="Xamarin.Build.Download" version="0.4.6" />
-        <dependency id="Xamarin.Firebase.iOS.InstanceID" version="2.0.8.0" />
-        <dependency id="Xamarin.Firebase.iOS.Core" version="4.0.13.1" />
+        <dependency id="Xamarin.Build.Download" version="0.4.11" />
+        <dependency id="Xamarin.Firebase.iOS.InstanceID" version="3.1.1.0" />
+        <dependency id="Xamarin.Firebase.iOS.Core" version="5.0.5.0" />
       </group>
     </dependencies>
   </metadata>
```
[the remark this answers is not in the document]
Comment (Redth) [reply Firebase.Analytics/nuget/Xamarin.Firebase.iOS.Analytics.nuspec]: Using XBD, not needed, nm.

File: Firebase.CrashReporting/nuget/Xamarin.Firebase.iOS.CrashReporting.nuspec [hunk 11]
```diff
@@ -14,10 +14,10 @@
...
-        <dependency id="Xamarin.Firebase.iOS.Analytics" version="4.0.5.1" />
-        <dependency id="Xamarin.Firebase.iOS.InstanceID" version="2.0.8.0" />
-        <dependency id="Xamarin.Firebase.iOS.Core" version="4.0.13.1" />
+        <dependency id="Xamarin.Build.Download" version="0.4.11" />
+        <dependency id="Xamarin.Firebase.iOS.Analytics" version="5.0.0.0" />
+        <dependency id="Xamarin.Firebase.iOS.InstanceID" version="3.0.0.0" />
+        <dependency id="Xamarin.Firebase.iOS.Core" version="5.0.5.0" />
       </group>
     </dependencies>
   </metadata>
```
[the remark this answers is not in the document]
Comment (Redth) [reply Firebase.CrashReporting/nuget/Xamarin.Firebase.iOS.CrashReporting.nuspec]: Using XBD, not needed, nm.
